### ISSUE TITLE
Tensorflow backend is 4x slower than Theano

### ISSUE THREAD

**Issue by elanmart** [fchollet/keras#2156]
@fchollet this should interest you.

Below you can find attached timings for 3 runs of the model with different backends. 
It's a simple FFN, 3 layers + dropouts + batchnorm (see the attached code for details)

It seems like TensorFlow backend is extremely inefficient and should not be used at the moment. 
Even with native TF code, Theano still comes on top. Perhaps this should be mentioned in the docs.

Perhaps 1.0 release will make it easier to work with TF itself and make the tf backend more useful?

`TensorFlow` is a standard `K.backend`
`TensorFlowNative` is a quick hack to replace everything from softmax up (softmax, optimizer, updates) with TF code
`Theano` is benched with and without `allow_gc`

GPU: `GTX 980` + `cuda-7.5` + `cudnn-4.0.7`

Backend | Timing
--- | ---
TensorFlow | `[29.6s, 29.8s, 30.6s]`
TensorFlowNative | `[11.6s, 11.4s, 11.5s]`
Theano noGC | `[7.4s, 7.6s, 7.4s]`
Theano GC | `[7.7s, 7.66s, 7.66s]`

Here's the script to reproduce those results.

```python
import numpy as np
import numpy.random as nr

from keras.utils import np_utils
from keras.layers.core import Layer, Dense, Dropout
from keras.layers.normalization import BatchNormalization
from keras.models import Sequential

X = np.random.randn(10000, 500)
y = np.random.randint(0, 64, size=(10000, 1))
y = np_utils.to_categorical(y)

test = Sequential()
test.add(Layer(input_shape=(500, )))

test.add(Dense(1024, activation='relu'))
test.add(Dropout(0.5))
test.add(BatchNormalization())

test.add(Dense(1024, activation='relu'))
test.add(Dropout(0.5))
test.add(BatchNormalization())

test.add(Dense(1024, activation='relu'))
test.add(Dropout(0.5))
test.add(BatchNormalization())

test.add(Dense(64, activation='softmax'))

test.compile('adam', 'categorical_crossentropy')

%%time 
test.fit(X, y, nb_epoch=25, verbose=0, )
```

**Comment by meanmee** [comment]
in my computer they are almost the same runtime

**Comment by elanmart** [comment]
@meanmee That's interesting. Are You using GPU?

**Comment by fluency03** [comment]
Is this including compiling time, or only training time?

**Comment by meanmee** [comment]
Yes， I use Titan x, And I noticed tensorflow's compiling time is faster than theano's, but sometimes tensorflow is very slow(compliing time), maybe it is slow the first time to compile tensorflow.

**Comment by fluency03** [comment]
Maybe it is also interesting that you could compare the training time and compiling time separately. Then the idea will be more clear.

**Comment by elanmart** [comment]
@fluency03 runtime only.
@meanmee this is super weird. Adding `tf.Session(config=tf.ConfigProto(log_device_placement=True))` confirms that my TF is using the GPU as well

**Comment by fchollet** [comment]
In my experience Theano and TF are roughly the same runtime (Theano usually ~5-10% faster) both on CPU and GPU, except for RNNs. For RNNs Theano is easily 3x faster.

**Comment by fchollet** [comment]
So currently I am not able to reproduce these results at all. Can you post your native TF code as well?

**Comment by fchollet** [comment]
PS: running your code, I'm seeing 7s/epoch for Theano and 16s/epoch for TF. If I change the optimizer to SGD (and change nothing else) I'm seeing 6s/epoch for Theano and 10s/epoch for TF. So it seems the actually issue here is "Adam is slow with TensorFlow". It is a reasonable point, but it's a far cry from the original claims...

**Comment by fchollet** [comment]
Ok, so, I retried with the latest version of Keras. Your code runs at 10s/epoch (with Adam) in TF and 7s/epoch with Theano. Sounds reasonable.

But then I tried implementing pure-TF training, and it ran at 6.5s/epoch. So it seems like the Keras way to run weight updates imposes a penalty on TF training, which doesn't exist with Theano. It's not all that surprising (you would expect the native way to run weight updates to be better optimized), but it in any case it warrants further investigation into how we can make things faster.

**Comment by fchollet** [comment]
An update on this.

I have implemented native TF optimizer training in Keras and benchmarked it against a variety of scripts. It's not very complicated, it consists in passing the optimizer, loss and constraints to K.function instead of just passing a list of updates, and handle weight updating in K.function, with the TF side optionally using a TF training op from a native TF optimizer instead of calling `get_updates` on the Keras optimizer. It's just a few lines.

**It does not result in a speed-up that would justify pushing such an update.** I could only observe a speed up in the case of your example, which seems to be an edge case. For several examples I am seeing *faster* runtime with Keras optimizers.

Runtime per epoch on CPU for a few scripts (native vs. keras):

- your example: **7s vs. 10s**. This is not a 30% speedup, this is a 3s fixed overhead. By doubling the number of modules in the model (module = Dense + BN + DP) you get 16s vs. 19s. Not a big deal tbh.
- imdb_cnn: **255s vs. 243s**: Keras optimizer is faster than native, although it may be just noise.
- imdb_lstm: **296 vs. 306s**: again, faster than native, although it may be just noise.
- mnist_mlp: **10s vs. 10s**

It's possible that your example is different because of BN. In any case, the difference is not consequential.

Conclusion: we won't push native TF optimizer based training.

**Comment by saulshanabrook** [comment]
@elanmart how did you [check if Tensorflow is using the GPU](https://github.com/fchollet/keras/issues/2156#issuecomment-204364380) with keras?

**Comment by kengz** [comment]
as a follow up, someone did a [benchmark comparing speed between TF and Theano](http://www.ccri.com/2016/12/09/torch-vs-tensorflow-vs-theano/). TLDR: Theano is consistently faster, and the difference can be indeed by x4 in some cases, or not much difference in other.

**Comment by jerpint** [comment]
Hello all, I am currently trying to use theano as my backend. Using the mnist_cnn.py as benchmark, I find that an epoch takes approx. 10 seconds with TF but about 50 seconds with Theano.  I am using GPU for both. 

This is what I have in my .theanorc file : 

```
[global]
floatX = float32
device = cuda


[dnn]
enabled = True
include_path = /usr/local/cuda/include
library_path = /usr/local/cuda/lib64
```

and this is what shows in terminal when I run : 

```
Using Theano backend.
Using cuDNN version 5110 on context None
Mapped name None to device cuda: GeForce GTX 1050 (0000:01:00.0)
x_train shape: (60000, 28, 28, 1)
60000 train samples
10000 test samples
...
```

when I use the CPU for theano, runtime is approx 250 seconds. This confirms that the GPU is being used, but theano is still running much slower, which according to this thread should not be happening.

Any ideas as to what may be causing this?

Thank you

**Comment by the-moliver** [comment]
Make sure you use 'channels_first' as the dim ordering for Theano, it will be much faster. Otherwise you have to do a lot of dim shuffling which slows things down.

**Comment by jerpint** [comment]
That helped quite a bit!

But now I am at 25 seconds vs 10 seconds, any other tips?

**Comment by the-moliver** [comment]
Check all your dimensions with model.summary() to make sure everything is behaving as it should, x_train shape should be (60000, 1, 28, 28), etc.

**Comment by jerpint** [comment]
everything seems to be in order with the model.summary() : 

```
Using Theano backend.
Using cuDNN version 5110 on context None
Mapped name None to device cuda: GeForce GTX 1050 (0000:01:00.0)
x_train shape: (60000, 1, 28, 28)
60000 train samples
10000 test samples
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
conv2d_1 (Conv2D)            (None, 32, 26, 26)        320       
_________________________________________________________________
conv2d_2 (Conv2D)            (None, 64, 24, 24)        18496     
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 64, 12, 12)        0         
_________________________________________________________________
dropout_1 (Dropout)          (None, 64, 12, 12)        0         
_________________________________________________________________
flatten_1 (Flatten)          (None, 9216)              0         
_________________________________________________________________
dense_1 (Dense)              (None, 128)               1179776   
_________________________________________________________________
dropout_2 (Dropout)          (None, 128)               0         
_________________________________________________________________
dense_2 (Dense)              (None, 10)                1290      
=================================================================
Total params: 1,199,882
Trainable params: 1,199,882
Non-trainable params: 0
_________________________________________________________________
Train on 60000 samples, validate on 10000 samples
Epoch 1/12
 8960/60000 [===>..........................] - ETA: 20s - loss: 0.9396 - acc: 0.7110

```

**Comment by the-moliver** [comment]
Can you post your .theanorc file

**Comment by jerpint** [comment]
```
[global]
floatX = float32
device = cuda


[dnn]
enabled = True
include_path = /usr/local/cuda/include
library_path = /usr/local/cuda/lib64
```

**Comment by the-moliver** [comment]
Hmmm. I would make sure you have latest drivers, CUDA and Theano. You could try preallocating memory:
```
[gpuarray]
preallocate = .75
```
Or experiment with the optimizer flags: http://deeplearning.net/software/theano/library/config.html#config.optimizer

But that really shouldn't be necessary...

**Comment by jerpint** [comment]
I have cuda 8.0 and just reinstalled the latest theano-dev version.  When I installed tensorlow, I built it from source using bazel and the optimal configuration based on my system.  When I had only pip installed tensorflow, I was getting warnings saying I could improve performance by doing so. Could the optimization on the tensorflow side of things account for the difference?  

https://www.tensorflow.org/install/install_sources

**Comment by davidtellez** [comment]
> Make sure you use 'channels_first' as the dim ordering for the Theano backend, it will be much faster. Otherwise Keras has to do a lot of dim shuffling which slows things down. I can see from your post you are using 'channels_last'; x_train shape should be (60000, 1, 28, 28) when used with 'channels_first'.

**TL;DR**: set image_data_format to "channels_first" if you are using the Theano backend, otherwise, you might experience slow downs (definitely happens with batch normalization). 

I tried training a network on Windows and Ubuntu with Keras+Theano using the channel order by default. According to keras.json, the default was "channels_last". I could train as usual and didn't experience anything weird. At some point, I added batch normalization and observed a massive slow down (training epochs taking 2-4 times more time). I tried many things without success until I saw @the-moliver comment above. **Matching the channel order to the backend you use is not optional but necessary (I think this is not clear in the documentation).** I'd advise to change the documentation or even the code to enforce the use of the channel order that matches the backend (tensorflow: channels_first, theano: channels_last).

Thanks @the-moliver!! :)
